presto-main/src/main/java/io/prestosql/operator/WorkProcessorSourceOperatorAdapter.java

```diff
@@ -0,0 +1,229 @@
+                    long currentInternalNetworkPositions = sourceOperator.getInternalNetworkPositions();
+
+                    long currentInputBytes = sourceOperator.getInputDataSize().toBytes();
+                    long currentInputPositions = sourceOperator.getInputPositions();
+
+                    if (currentPhysicalInputBytes != previousPhysicalInputBytes
+                            || currentPhysicalInputPositions != previousPhysicalInputPositions
+                            || currentReadTimeNanos != previousReadTimeNanos) {
+                        operatorContext.recordPhysicalInputWithTiming(
+                                currentPhysicalInputBytes - previousPhysicalInputBytes,
+                                currentPhysicalInputPositions - previousPhysicalInputPositions,
+                                currentReadTimeNanos - previousReadTimeNanos);
+
+                        previousPhysicalInputBytes = currentPhysicalInputBytes;
+                        previousPhysicalInputPositions = currentPhysicalInputPositions;
+                        previousReadTimeNanos = currentReadTimeNanos;
+                    }
+
+                    if (currentInternalNetworkInputBytes != previousInternalNetworkInputBytes
+                            || currentInternalNetworkPositions != previousInternalNetworkPositions) {
+                        operatorContext.recordNetworkInput(
+                                currentInternalNetworkInputBytes - previousInternalNetworkInputBytes,
+                                currentInternalNetworkPositions - previousInternalNetworkPositions);
+
+                        previousInternalNetworkInputBytes = currentInternalNetworkInputBytes;
+                        previousInternalNetworkPositions = currentInternalNetworkPositions;
+                    }
+
+                    if (currentInputBytes != previousInputBytes
+                            || currentInputPositions != previousInputPositions) {
+                        operatorContext.recordProcessedInput(
+                                currentInputBytes - previousInputBytes,
+                                currentInputPositions - previousInputPositions);
+
+                        previousInputBytes = currentInputBytes;
+                        previousInputPositions = currentInputPositions;
+                    }
+                })
+                .finishWhen(() -> operatorFinishing);
+    }
+
+    @Override
+    public PlanNodeId getSourceId()
+    {
+        return sourceId;
+    }
+
+    @Override
+    public Supplier<Optional<UpdatablePageSource>> addSplit(Split split)
+    {
+        if (operatorFinishing) {
+            return Optional::empty;
+        }
+
+        Object splitInfo = split.getInfo();
+        if (splitInfo != null) {
+            operatorContext.setInfoSupplier(() -> new SplitOperatorInfo(splitInfo));
+        }
+
+        pendingSplits.add(split);
+        blockedOnSplits.set(null);
```

Comment:
This can be addressed by modeling `SplitBuffer` similarly as `WorkProcessorOperatorAdapter.PageBuffer`. `SplitBuffer` would be passed to `WorkProcessorSourceOperatorFactory#create` instead of `WorkProcessor<Split>` as it's currently. Source operator would be able to register split listener so that adding splits is independent process of producing pages (as in `ExchangeOperator`).